src/Application.php

```diff
@@ -1,112 +1,133 @@
 <?php
 /**
  * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
- * @copyright Copyright (c) 2015-2017 Zend Technologies USA Inc. (https://www.zend.com)
+ * @copyright Copyright (c) 2018 Zend Technologies USA Inc. (https://www.zend.com)
  * @license   https://github.com/zendframework/zend-expressive/blob/master/LICENSE.md New BSD License
  */
 
 declare(strict_types=1);
 
 namespace Zend\Expressive;
 
-use Fig\Http\Message\StatusCodeInterface as StatusCode;
-use Interop\Http\Server\MiddlewareInterface;
-use Interop\Http\Server\RequestHandlerInterface;
-use InvalidArgumentException;
-use Psr\Container\ContainerInterface;
 use Psr\Http\Message\ResponseInterface;
 use Psr\Http\Message\ServerRequestInterface;
-use Throwable;
-use UnexpectedValueException;
-use Zend\Diactoros\Response;
-use Zend\Diactoros\Response\EmitterInterface;
-use Zend\Diactoros\Response\SapiEmitter;
-use Zend\Diactoros\ServerRequest;
-use Zend\Diactoros\ServerRequestFactory;
-use Zend\Stratigility\MiddlewarePipe;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Zend\Stratigility\MiddlewarePipeInterface;
 
-/**
- * Middleware application providing routing based on paths and HTTP methods.
- */
-class Application extends MiddlewarePipe
+use function Zend\Stratigility\path;
+
+class Application implements MiddlewareInterface, RequestHandlerInterface
 {
     use ApplicationConfigInjectionTrait;
-    use MarshalMiddlewareTrait;
-
-    const DISPATCH_MIDDLEWARE = 'EXPRESSIVE_DISPATCH_MIDDLEWARE';
-    const ROUTING_MIDDLEWARE = 'EXPRESSIVE_ROUTING_MIDDLEWARE';
 
     /**
-     * @var null|ContainerInterface
+     * @var MiddlewareFactory
      */
-    private $container;
+    private $factory;
 
     /**
-     * @var null|RequestHandlerInterface
+     * @var MiddlewarePipeInterface
      */
-    private $defaultDelegate;
+    private $pipeline;
 
     /**
-     * @var bool Flag indicating whether or not the dispatch middleware is
-     *     registered in the middleware pipeline.
+     * @var Middleware\RouteMiddleware
      */
-    private $dispatchMiddlewareIsRegistered = false;
+    private $routes;
 
     /**
-     * @var null|EmitterInterface
+     * @var ApplicationRunner
      */
-    private $emitter;
+    private $runner;
+
+    public function __construct(
+        MiddlewareFactory $factory,
+        MiddlewarePipeInterface $pipeline,
+        Middleware\RouteMiddleware $routes,
+        ApplicationRunner $runner
```

Comment:
This makes Application and ApplicationRunner a circular dependency

Reply:
No, it doesn't; the `MiddlewarePipe` is what is passed to the runner in this case.

Reply:
The circular dependency is due to how the factories are constructed; the `ApplicationRunner` pulls the `Application` service, but the `Application` service pulls the `ApplicationRunner`.

I'll figure out a solution to this shortly and push the changes.

Reply:
Fixed with 38c6ff5